source/extensions/filters/http/cache/cache_entry_utils.cc

```diff
@@ -28,6 +28,58 @@ std::ostream& operator<<(std::ostream& os, CacheEntryStatus status) {
   return os << cacheEntryStatusString(status);
 }
 
+namespace {
+const absl::flat_hash_set<Http::LowerCaseString> headersNotToUpdate() {
+  CONSTRUCT_ON_FIRST_USE(
+      absl::flat_hash_set<Http::LowerCaseString>,
+      // Content range should not be changed upon validation
+      Http::Headers::get().ContentRange,
+
+      // Headers that describe the body content should never be updated.
+      Http::Headers::get().ContentLength,
+
+      // It does not make sense for this level of the code to be updating the ETag, when
+      // presumably the cached_response_headers reflect this specific ETag.
+      Http::CustomHeaders::get().Etag,
+
+      // We don't update the cached response on a Vary; we just delete it
+      // entirely. So don't bother copying over the Vary header.
+      Http::CustomHeaders::get().Vary);
+}
+} // namespace
+
+void applyHeaderUpdate(const Http::ResponseHeaderMap& new_headers,
+                       Http::ResponseHeaderMap& headers_to_update) {
+  // Assumptions:
+  // 1. The internet is fast, i.e. we get the result as soon as the server sends it.
+  //    Race conditions would not be possible because we are always processing up-to-date data.
+  // 2. No key collision for etag. Therefore, if etag matches it's the same resource.
+  // 3. Backend is correct. etag is being used as a unique identifier to the resource
+
+  // use other header fields provided in the new response to replace all instances
+  // of the corresponding header fields in the stored response
+
+  // `updatedHeaderFields` makes sure each field is only removed when we update the header
+  // field for the first time to handle the case where incoming headers have repeated values
+  absl::flat_hash_set<Http::LowerCaseString> updatedHeaderFields;
+  new_headers.iterate(
+      [&headers_to_update, &updatedHeaderFields](
+          const Http::HeaderEntry& incoming_response_header) -> Http::HeaderMap::Iterate {
+        Http::LowerCaseString lower_case_key{incoming_response_header.key().getStringView()};
+        absl::string_view incoming_value{incoming_response_header.value().getStringView()};
+        if (headersNotToUpdate().contains(lower_case_key)) {
```

Comment:
perf nit: capture the map returned by headersNotToUpdate() outside, rather than calling it in the loop, as it has to do some kind of thread-safe access to lazy-init the map.

Reply:
Ah, I guess I shouldn't click "resolve conversation" as it apparently prevents you from seeing my replies! This bit is in PR #23977, and is moving a block of existing code that I didn't write. As such I'd appreciate if any cleanup of that code could be separate from moving it, and also separate from this whole chain of PRs.

Reply:
sure that makes sense; github review does not make it easy for me to tell when code is moved vs written. Sometimes when writing PRs I call out moves as comments I add to my own PR but they can get lost with new commits.

+1 to letting the initiator of a comment resolve it (most of the time). Sometimes if there are non-controversial suggestions that I just execute as indicated in the PR I'll  go ahead and resolve myself, but if it's just a response in a dialog it makes more sense to let the initiator resolve (IMO).

Reply:
I'm slowly getting my head around the github comment system - that the interface you see first won't let you post comments grouped, and that 'resolve' pseudo-disappears it immediately (I'm used the Google review format where a resolve-like button means "mark as resolved" and doesn't disappear anything - Github not keeping track of whether there's any comments I'm still supposed to be addressing is a pain!)